Catch MQTT payload encoding errros

hartmms opened the pull request:
**Description:**
When MQTT payloads are received by the MQTT component that are not utf-8 uincode, the component throws an exception and terminates.  Hass still runs, but is unable to do anything over MQTT protocol. 

This change added an exception handler to this situation and generates a logfile error message with details. It then does not post anything to the mqtt event bus. 

**Related issue (if applicable):** 

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#

**Example entry for `configuration.yaml` (if applicable):**

**Checklist:**

hartmms committed:
Added exception/error handling for when a MQTT payload unicode "decode" fails.

hartmms committed:
fixed formatting of mqtt utf-8 changes

kellerza committed:
Add voluptuous to locative (#3254)

turbokongen committed:
Revert only add 1 device (#3324)

Teagan42 committed:
Automatic Device Tracker Bug Fix (#3330)

* Iterate over items

* Pass display name as host name

balloob committed:
Comment out pyuserinput in requirements_all (#3307)

* Comment out pyuserinput in requirements_all

* Ignore import error for keyboard component

balloob committed:
Bugfix group order (#3323)

* Add ordered dict config validator

* Have group component use ordered dict config validator

* Improve config_validation testing

* update doc string config_validation.ordered_dict

* validate full dict entries

* Further simplify ordered_dict validator.

* Lint fix

clach04 committed:
bluetooth_le_tracker clarify header with LE (#3328)

Khabi committed:
Slack notification optional username / icon (#3314)

* Slack notification optional username / icon

* Small bugfix to handle defaults better

* Dedup'ing code

fabaff committed:
Update ordering, constants, and callback name (light.*) (#3339)

* Extend schema

* Update ordering

* Add line breaks

* Align callback name with other platforms

* ALign callbackname with other platforms

* Update callback name to match other platforms

* Update callback name

* Update callback name

fabaff committed:
Fix typo (#3343)

mKeRix committed:
Minor naming fix (#3318)

turbokongen committed:
Bugfix ecobee: inverted high and low temps and enforce int to temps (#3325)

* inverted high and low temps

* Looks like somethings are mixed up

* Added debugentires

* Added debugentires 2

* Enforce int on temperatures

pvizeli committed:
Update ha-ffmpeg version 0.12 and add tests (#3301)

* update ha-ffmpeg version 0.12 and add tests

* change error logging

pvizeli committed:
Bugfix voluptuous on recorder (#3350)

turbokongen committed:
Missing garage door detection (#3349)

balloob commented:
Can you add a test for this.

DavidLP committed:
Bugfix pilight component (#3355)

* BUG Message data cannot be changed thus use voluptuous to ensure format

* Pilight daemon expects JSON serializable data

Thus dict is needed and not a mapping proxy.

* Add explanation why dict as message data is needed

* Use more obvious voluptuous validation scheme

* Pylint:  Trailing whitespace

Commit:
Add open/closed state for open_cover and close_cover in SERVICE_TO_STATE (#3180)

* Add open/closed state mapping for open_cover and close_cover

* Add 'open', 'closed' for open/close_cover_tilt

* Revert "Add 'open', 'closed' for open/close_cover_tilt"

This reverts commit e45582d4394a33feedfce190a1dba96473d24825.

balloob committed:
Migrate core from threads to async awesomeness (#3248)

* Add event loop to the core

* Add block_till_done to HA core object

* Fix some tests

* Linting core

* Fix statemachine tests

* Core test fixes

* fix block_till_done to wait for loop and queue to empty

* fix test_core for passing, and correct start/stop/block_till_done

* Fix remote tests

* Fix tests: block_till_done

* Fix linting

* Fix more tests

* Fix final linting

* Fix remote test

* remove unnecessary import

* reduce sleep to avoid slowing down the tests excessively

* fix remaining tests to wait for non-threadsafe operations

* Add async_ doc strings for event loop / coroutine info

* Fix command line test to block for the right timeout

* Fix py3.4.2 loop var access

* Fix SERVICE_CALL_LIMIT being in effect for other tests

* Fix lint errors

* Fix lint error with proper placement

* Fix slave start to not start a timer

* Add asyncio compatible listeners.

* Increase min Python version to 3.4.2

* Move async backports to util

* Add backported async tests

* Fix linting

* Simplify Python version check

* Fix lint

* Remove unneeded try/except and queue listener appproriately.

* Fix tuple vs. list unorderable error on version compare.

* Fix version tests

hartmms commented:
I took a look a the tests/components/mqtt directory.  While I'm somewhat proficient at python script hacking, I'm not particularly sure how to write a test for this. I can see how to publish to the test-topic, but I don't know what the right self.assertTrue/assertFalse/assertEqual "test" would be.  

Am I in the right ballpark with something like this?

```python
def test_publish_calls_service(self):
  """Test the mqtt component still lives after a non utf-8 payload received."""
  mqtt.publish(self.hass, 'test-topic', 0x9a)
  self.assertTrue(mqtt.MQTT_CLIENT.progress) 
```

Sorry I'm not more versed in test case writing.

beepmill committed:
Ignore desktop.ini (Windows Explorer) (#3363)

Teagan42 committed:
Automatic polling (#3360)

* Test updating automatic

* Scan interval

* Schedule scan every time delta

* Pass around has

* Recursive issue

* Method invocation

* Oops

* Set up poll

* Default argument value

* Unused import

* Semicolon

* Fix tests

* Linting

* Unneeded throttle as it's handled by time event

* Use track time change event listener

* Disable lint rule

* Attribute removed - removing test

* Debug instead of info

* Unused import

balloob commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -393,13 +393,17 @@ def _mqtt_on_subscribe(self, _mqttc, _userdata, mid, granted_qos):
+            })
+        except UnicodeDecodeError:
+            _LOGGER.error("Illegal utf-8 unicode payload from "
+                          "MQTT topic %s. Payload: %s", msg.topic, msg.paylad)
```
`msg.payload`

balloob commented:
Try something like this:

```python
    def test_receiving_non_utf8_message_gets_logged(self, caplog):
        calls = []

        def record(event):
            """Helper to record calls."""
            calls.append(event)

        self.hass.bus.listen_once(mqtt.EVENT_MQTT_MESSAGE_RECEIVED, record)

        MQTTMessage = namedtuple('MQTTMessage', ['topic', 'qos', 'payload'])
        message = MQTTMessage('test_topic', 1, 0x9a)

        mqtt.MQTT_CLIENT._mqtt_on_message(None, {'hass': self.hass}, message)
        self.hass.block_till_done()

        assert len(calls) == 0
        record = caplog.records[-1]
        assert record.levelname == 'ERROR'
        assert record.msg == 'Illegal utf-8 unicode payload from MQTT topic test_topic.'
```

fabaff committed:
Use voluptuous for SCSGate (#3265)

* Migrate to voluptuous

* Extend platforms

turbokongen committed:
Fix temp conversion for nest setpoint (#3373)

* Fix temp conversion for nest setpoint

* DEbug

fabaff committed:
Add Simplepush notifications (#3336)

turbokongen committed:
zxt_120 set temperature did not update on setpoint (#3380)

pvizeli committed:
fix xbox live entity id (#3368)

pvizeli committed:
Bugfix auto/manual mode change (#3384)

persandstrom committed:
Modbus component refactoring - sensors and switches (#3297)

ericwclymer committed:
Fix saving push_notification.conf as suggested by @robbiet480 (#3382)

pvizeli committed:
SleepIQ component with sensor and binary sensor platforms (#3390)

Original from #2949

wokar committed:
fixed link constructor to show icon again (#3388)

kellerza committed:
Fix tests (#3387)

daBONDi committed:
Implement Sensor for KNX Platform (#2911)

* Added Sensor Support for KNX Devices

Added Sensor for KNX Group Addresses
- Temperature
- Wind Speed
- Illuminance(LUX)

Mostly to fetch from a KNX Wetterstation

* Some pylint,flake8 fixes

* Pydoc Fixes

* Fix Coverage Ordering

* Refactor KNX Sensor

Refactor to Idea from @usul27 and added Minimum Maximum

* Removed Measurement Untis from const.py

Removed needed Measurement Units from const.py and add it to
sensor\knx.py

* Change .coveragerc

* Add as Requested from @Teagan42 the new Type Names

Additional add CONF_MINIMUM and CONF_MAXIMUM

* Added Changes as Requested from @Teagan42

* Fixed the Merge Conflict, Hopefully i done it rigth :-)

* Fixed Styling

fabaff committed:
Use voluptuous for statsd (#2928)

* Migrate to voluptuous

* Update tests

mKeRix committed:
Added away timeout setting for idle devices (#3321)

robjohnson189 committed:
Vera Thermostat Support (#3247)

* vera thermostat & climate support

* disable abstract

* code review changes

* fix

* fix build

* remove old method

fabaff committed:
Use voluptuous for Yamaha receiver (#3210)

* Migrate to voluptuous

* Add missing configuration variables

nkgilley committed:
short sleep (#3115)

Commit:
Miflora (#3053)

* First version of the MiFlora sensor (not yet finished)

* First workign version

* Added some documentation
Get name from sensor, if not defined

* Ignore IOError

* Added force_update option

* Updated comments

* Renamed fertility to conductivity (what it really is)

* MiFlora library update

* Updated helper files

* Formatting

* Fixed pylint errors

* Removed default from monitored conditions

* Removed KeyError handling as a KeyError should never be raised

* Added a return when no data is received

* emoved unnecessary return statement

* Changed default name

* Changes quotes and string operation ( @Teagan42 )

* - number of samples for median calculation is now configurable
- set state to None if no data could be polled from sensor

* Bugfix in library
more logging

* Fixed miflora version number

fabaff committed:
Upgrade sqlalchemy to 1.0.15 (#3333)

fabaff committed:
Upgrade cherrypy to 8.1.0 (#3334)

fabaff committed:
Upgrade PyMata to 2.13 (#3335)

fabaff committed:
Migrate to voluptuous (#3337)

fabaff committed:
Use voluptuous for iTunes (#3344)

* Migrate to voluptuous

* Add support for SSL/TLS

fabaff committed:
Use voluptuous for KNX (#3345)

* Migrate to voluptuous

* Make host optional and set default

pvizeli committed:
Update yahooweather 0.8 / change request time (#3352) [Breaking change]

jawilson committed:
Fix octoprint sensor (#3385)

* Fix non-temperature sensors for octoprint

* Fix double space in octoprint temperature names

* Fix tox linting errors

fabaff committed:
Use voluptuous for Zone (#3377)

* Migrate to voluptuous

* Zone: Remove unneeded latitude/longitude check

fabaff committed:
Migrate to voluptuous (#3372)

fabaff committed:
Use voluptuous for Tellstick (#3367)

* Migrate to voluptuous

* Update tellstick.py

Commit:
fix style in tellstick sensor

chrom3 committed:
Kodi notification platform (#3403)

ej81 committed:
Bugfix: incorrectly inverted value when setting cover position (#3376)

especially-relative committed:
Add operation_list to radiotherm (#3393)

* Add operation_list to radiotherm

* Use constants

Commit:
Nuimo controller component based on SDK (#3039)

especially-relative committed:
Implemented onkyo reconnect (#3061)

* Implemented onkyo reconnect

Connection object is cleared after a failed command. It will be
automatically recreated upon the next command running. This should allow
for failed connections to be restored.

* Remove reduntant error catching

* Run all update commands with command wrapper.

* Handle errors better

* Removed unused global

I have no idea how that got there.

pvizeli committed:
Split ffmpeg to compoment (#3396)

Add an optional extended description…

pavoni committed:
Add support for Vera covers. (#3411)

balloob committed:
Fix zones (#3413)

jawilson committed:
Add ability to set fan made to Nest climate component (#3399)

* Add ability to set fan made to Nest climate component

* Use constants for fan values

* Use STATE_ON from cost

* Fix lint error

balloob committed:
Use setup_component in tests (#3414)

* Alarm Control Panel Manual - use setup_component

* Update automation - zone tests

* Update climate - demo tests

* Update climate - generic thermostat tests

* Update cover - command line tests

* Update cover - demo tests

* Update device tracker tests

* Update device tracker - owntracks tests

* Update fan - demo tests

* Update garage door - demo tests

* Update light tests

* Update lock - demo tests

* Update media player - demo tests

* Update notify - command line tests

* Update notify - demo tests

* Update notify - file tests

* Update notify - group tests

* Update sensor - mfi tests

* Update sensor - moldindicator tests

* Update sensor - mqtt room tests

* Update switch - command line

* Update switch - flux

* Update switch tests

* Update scene tests

* Fix wrong default port for mfi switch

hartmms committed:
MQTT component:
  added Attribute exception in case the mqtt message is not seen as encoded unicode

  added test

hartmms committed:
MQTT component:
   remove on_message_received Attribute exception handler because that will never happen

balloob commented:
This PR seems to have gone stale. Closing it. You can reopen it when you're ready to finish it. (and you'll have to clean up the branch too)
